source/particle/generator/ascii_file.cc

```diff
@@ -0,0 +1,191 @@
+/*
+  Copyright (C) 2011 - 2015 by the authors of the ASPECT code.
+
+ This file is part of ASPECT.
+
+ ASPECT is free software; you can redistribute it and/or modify
+ it under the terms of the GNU General Public License as published by
+ the Free Software Foundation; either version 2, or (at your option)
+ any later version.
+
+ ASPECT is distributed in the hope that it will be useful,
+ but WITHOUT ANY WARRANTY; without even the implied warranty of
+ MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+ GNU General Public License for more details.
+
+ You should have received a copy of the GNU General Public License
+ along with ASPECT; see the file doc/COPYING.  If not see
+ <http://www.gnu.org/licenses/>.
+ */
+
+#include <aspect/particle/generator/ascii_file.h>
+
+#include <deal.II/grid/grid_tools.h>
+
+
+namespace aspect
+{
+  namespace Particle
+  {
+    namespace Generator
+    {
+      template <int dim>
+      AsciiFile<dim>::AsciiFile() {}
+
+      template <int dim>
+      void
+      AsciiFile<dim>::generate_particles(World<dim> &world)
+      {
+        const std::string filename = data_directory+data_filename;
+        std::ifstream in(filename.c_str(), std::ios::in);
+        AssertThrow (in,
+                     ExcMessage (std::string("Couldn't open data file <"
+                                             +
+                                             filename
+                                             +
+                                             ">.")));
+
+        // Read header lines and table size
+        while (in.peek() == '#')
+          {
+            std::string temp;
+            getline(in,temp);
+          }
+
+        // Read data lines
+        particle_index id = 0;
+        Point<dim> coordinates;
+
+        if (dim == 2)
+          while (in >> coordinates[0] >> coordinates[1])
+            {
+              generate_particle(coordinates,id,world);
+              id++;
+            }
+        else if (dim == 3)
+          while (in >> coordinates[0] >> coordinates[1] >> coordinates[2])
+            {
+              generate_particle(coordinates,id,world);
+              id++;
+            }
+        else
+          Assert(false,ExcNotImplemented());
+      }
+
+
+
+      template <int dim>
+      void
+      AsciiFile<dim>::generate_particle(const Point<dim> &position,
+                                        const particle_index id,
+                                        World<dim> &world)
```

Comment:
I think you probably have this function (or minor variations of it) in a number of .cc files. It may be worthwhile actually putting this into the base class instead to avoid the duplication.